scripts/validate_rst_title_capitalization.py

```diff
@@ -0,0 +1,214 @@
+./scripts/validate_rst_title_capitalization.py doc/source/
+
+"""
+import argparse
+import sys
+import re
+import os
+from typing import Tuple, Generator, List
+import glob
+
+
+CAPITALIZATION_EXCEPTIONS = {
+    "pandas",
+    "Python",
+    "IPython",
+    "PyTables",
+    "Excel",
+    "JSON",
+    "HTML",
+    "SAS",
+    "SQL",
+    "BigQuery",
+    "STATA",
+    "Interval",
+    "PEP8",
+    "Period",
+    "Series",
+    "Index",
+    "DataFrame",
+    "C",
+    "Git",
+    "GitHub",
+    "NumPy",
+    "Apache",
+    "Arrow",
+    "Parquet",
+    "MultiIndex",
+    "NumFOCUS",
+    "sklearn",
+    "Docker",
+}
+
+CAP_EXCEPTIONS_DICT = {word.lower(): word for word in CAPITALIZATION_EXCEPTIONS}
+
+err_msg = "Heading capitalization formatted incorrectly. Please correctly capitalize"
+
+
+def correct_title_capitalization(title: str) -> str:
+    """
+    Algorithm to create the correct capitalization for a given title.
+
+    Parameters
+    ----------
+    title : str
+        Heading string to correct.
+
+    Returns
+    -------
+    correct_title : str
+        Correctly capitalized heading.
+
+    """
+
+    # Function to strip all non-word characters from the beginning of the title to the
+    # first word character.
+    correct_title: str = re.sub(r"^\W*", "", title).capitalize()
+
+    # Function to remove a URL from the title. We do this because words in a URL must
+    # stay lowercase, even if they are a capitalization exception.
+    removed_https_title = re.sub(r"<https?:\/\/.*[\r\n]*>", "", correct_title)
+
+    # Function to split a title into a list using non-word character delimiters.
+    word_list = re.split(r"\W", removed_https_title)
+
+    for word in word_list:
+        if word.lower() in CAP_EXCEPTIONS_DICT:
+            correct_title = re.sub(
+                rf"\b{word}\b", CAP_EXCEPTIONS_DICT[word.lower()], correct_title
+            )
+
+    return correct_title
+
+
+def find_titles(rst_file: str) -> Generator[Tuple[str, int], None, None]:
+    """
+    Algorithm to identify particular text that should be considered headings in an
+    RST file.
+
+    See <https://thomas-cokelaer.info/tutorials/sphinx/rest_syntax.html> for details
+    on what constitutes a string as a heading in RST.
+
+    Parameters
+    ----------
+    rst_file : str
+        RST file to scan through for headings.
+
+    Yields
+    -------
+    title : str
+        A heading found in the rst file.
+
+    line_number : int
+        The corresponding line number of the heading.
+
+    """
+
+    with open(rst_file, "r") as file_obj:
+        lines = file_obj.read().split("\n")
+
+    symbols = ("*", "=", "-", "^", "~", "#", '"')
+
+    table = str.maketrans("", "", "*`_")
+
+    for i, line in enumerate(lines):
+        if len(line) != 0 and len(lines[i - 1]) != 0:
+            line_chars = set(line)
+            if len(line_chars) == 1 and line_chars.pop() in symbols:
```

Comment:
Why is not as simple as checking that besides this condition the line and the previous have the same length?

I guess I'm missing something, but can't see why all next lines are required, sorry.

Reply:
I read the RST documentation guide on what is considered a heading, and there are two situations. 
Case 1:
The heading is between two lines like so:
*******
heading
*******
In this situation, the heading has the be the same length as the length of the line below it.

Case 2:
Heading
***************
In this case, there are no stars above the heading.  In this situation, the heading length only needs to be less than or equal to the length of the line below it to be considered a heading in rST

Reply:
I think we always use case 2, with the same length in both lines. I'd personally ignore all options of rst and make the script simpler. What do you think?

Reply:
Ok that sounds good to me.  I'll make the changes soon

Reply:
@datapythonista made the appropriate changes.